### PR TITLE
Modcat 214 2

### PR DESCRIPTION
**Pull request by eturienzo**
## Purpose

<a href="https://issues.folio.org/browse/MODCAT-214">MODCAT-214</a> Create a template for authority records
## Learning

<ul>
<li> Get authority template:</li>
<ul>
<li>Get the template from database</li>
<li>Check the mandatory tags</li>
<li>Creation an authority empty record from template </li>
</ul>
<li>Unit tests updates:</li>
<ul>
    <li>Units tests have been updated for get an empty record from template</li>
  </ul>
</ul>

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -260,6 +260,16 @@
           ],
           "permissionsRequired": []
         },
+        {
+          "methods": [
+            "GET"
+          ],
+          "pathPattern": "/marccat/authority-record/from-template/{idTemplate}",
+          "modulePermissions": [
+            "configuration.all"
```
Comment (carmentrazza) [inline descriptors/ModuleDescriptor-template.json:269]: Change configuration.all permissions

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.folio.marccat.dao.AbstractDAO;
+import org.folio.marccat.dao.AuthorityModelDAO;
+
+/**
+ * @author elena
+ *
+ */
+public class AuthorityModel extends Model {
+  private String frbrSecondGroup;
+  private String frbrThirdGroup;
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java:12]: Remove these two properties

File: src/main/java/org/folio/marccat/resources/AuthorityRecordAPI.java [hunk 3]
```diff
@@ -63,4 +67,20 @@
 
   }
 
+  @GetMapping("/authority-record/from-template/{idTemplate}")
+  public AuthorityRecord getEmptyRecord(@PathVariable final Integer idTemplate, @RequestParam final String lang,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/AuthorityRecordAPI.java:71]: You can format with one annotation per line?

File: src/main/java/org/folio/marccat/resources/BaseResource.java [hunk 4]
```diff
@@ -34,5 +48,68 @@
     return new ResponseEntity<>(errors, HttpStatus.NOT_FOUND);
   }
 
+  public void getEmptyRecord(final boolean isBib, final Record record, final RecordTemplate template, final String lang,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BaseResource.java:51]: You can move the method to the RecordUtils class

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 5]
```diff
@@ -1,132 +1,99 @@
...
   }
 
   @GetMapping("/bibliographic-record/from-template/{idTemplate}")
-  public BibliographicRecord getEmptyRecord(
-    @PathVariable final Integer idTemplate,
-    @RequestParam final String lang,
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public BibliographicRecord getEmptyRecord(@PathVariable final Integer idTemplate, @RequestParam final String lang,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:73]: Format w/ one annotation per line.

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 6]
```diff
@@ -1,132 +1,99 @@
...
 
-
   @PostMapping("/bibliographic-record")
-  public ResponseEntity<Object> save(
-    @RequestBody final ContainerRecordTemplate container,
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestParam final String lang,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public ResponseEntity<Object> save(@RequestBody final ContainerRecordTemplate container,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:86]: Format w/ one annotation per line.

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 7]
```diff
@@ -1,132 +1,99 @@
...
 public class BibliographicRecordAPI extends BaseResource {
 
-
   @GetMapping("/bibliographic-record/{id}")
-  public ResponseEntity<Object> getRecord(
-    @PathVariable final Integer id,
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public ResponseEntity<Object> getRecord(@PathVariable final Integer id,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:55]: Format w/ one annotation per line.

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 8]
```diff
@@ -1,132 +1,99 @@
...
-    @PathVariable final Integer idTemplate,
-    @RequestParam final String lang,
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public BibliographicRecord getEmptyRecord(@PathVariable final Integer idTemplate, @RequestParam final String lang,
+      @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
+      @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
+      @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
     return doGet((storageService, configuration) -> {
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:77]: Align the line

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 9]
```diff
@@ -1,132 +1,99 @@
...
-  public ResponseEntity<Object> getRecord(
-    @PathVariable final Integer id,
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public ResponseEntity<Object> getRecord(@PathVariable final Integer id,
+      @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
+      @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
+      @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
     return doGet((storageService, configuration) -> {
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:59]: Align the line

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 10]
```diff
@@ -1,132 +1,99 @@
...
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestParam final String lang,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public ResponseEntity<Object> save(@RequestBody final ContainerRecordTemplate container,
+      @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
+      @RequestParam final String lang, @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
+      @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
 
     return doPost((storageService, configuration) -> {
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:91]: Align the line

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 11]
```diff
@@ -185,12 +156,10 @@ public void delete(@PathVariable final String id,
...
   @ResponseStatus(HttpStatus.NO_CONTENT)
   @DeleteMapping("/bibliographic-record/unlock/{id}")
-  public void unlock(@PathVariable final String id,
-                     @RequestParam final String uuid,
-                     @RequestParam final String userName,
-                     @RequestParam final LockEntityType type,
-                     @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-                     @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public void unlock(@PathVariable final String id, @RequestParam final String uuid,
+      @RequestParam final String userName, @RequestParam final LockEntityType type,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:160]: Format w/ one annotation per line.

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 12]
```diff
@@ -200,31 +169,25 @@ public void unlock(@PathVariable final String id,
...
 
-
   @PutMapping("/bibliographic-record/lock/{id}")
-  public void lock(@PathVariable final String id,
-                   @RequestParam final String uuid,
-                   @RequestParam final String userName,
-                   @RequestParam final LockEntityType type,
-                   @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-                   @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public void lock(@PathVariable final String id, @RequestParam final String uuid, @RequestParam final String userName,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:173]: Fformat w/ one annotation per line.

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java [hunk 13]
```diff
@@ -200,31 +169,25 @@ public void unlock(@PathVariable final String id,
...
 
-
   @GetMapping("/bibliographic-record/duplicate")
-  public ResponseEntity<Object> duplicate(
-    @RequestParam final Integer id,
-    @RequestParam final String lang,
-    @RequestParam(name = "view", defaultValue = View.DEFAULT_BIBLIOGRAPHIC_VIEW_AS_STRING) final int view,
-    @RequestHeader(Global.OKAPI_TENANT_HEADER_NAME) final String tenant,
-    @RequestHeader(Global.OKAPI_URL) String okapiUrl) {
+  public ResponseEntity<Object> duplicate(@RequestParam final Integer id, @RequestParam final String lang,
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:187]: Format w/ one annotation per line and align

File: src/main/java/org/folio/marccat/resources/shared/RecordUtils.java [hunk 14]
```diff
@@ -25,27 +30,25 @@ private RecordUtils() {
...
-      Global.BIBLIOGRAPHIC_LEVEL_CODE +
-      Global.CONTROL_TYPE_CODE +
-      Global.CHARACTER_CODING_SCHEME_CODE +
-      Global.FIXED_LEADER_BASE_ADDRESS +
-      Global.ENCODING_LEVEL +
-      Global.DESCRIPTIVE_CATALOGUING_CODE +
-      Global.LINKED_RECORD_CODE +
-      Global.FIXED_LEADER_PORTION;
-  }
+  public static String getLeaderValue(boolean isBib) {
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/shared/RecordUtils.java:33]: It is preferable to align as before

File: src/main/java/org/folio/marccat/resources/shared/RecordUtils.java [hunk 15]
```diff
@@ -81,7 +84,8 @@ public static Field addTagTransactionDate(final String lang, final StorageServic
     fixed005.setCode(Global.DATETIME_TRANSACTION_TAG_CODE);
     fixed005.setDisplayValue(F.getFormattedToday("yyyyMMddHHmmss."));
     fixed005.setCategoryCode(Global.INT_CATEGORY);
-    fixed005.setDescription(storageService.getHeadingTypeDescription(Global.DATETIME_TRANSACTION_HEADER_TYPE, lang, Global.INT_CATEGORY));
+    fixed005.setDescription(
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/resources/shared/RecordUtils.java:87]: Put back on the same line

File: src/main/resources/hibernate.cfg.xml [hunk 16]
```diff
@@ -6,6 +6,8 @@
 	<session-factory>
 
 	<property name="show_sql">false</property>
+    <mapping resource="org/folio/marccat/dao/persistence/AUTHORITY_MODEL.hbm.xml"/>
+    <mapping resource="org/folio/marccat/dao/persistence/AUTHORITY_MODEL_ITEM.hbm.xml"/>
 		<mapping resource="org/folio/marccat/dao/persistence/BibliographicModel.hbm.xml"/>
```
Comment (carmentrazza) [inline src/main/resources/hibernate.cfg.xml:11]: Align the lines

File: src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java [hunk 17]
```diff
@@ -22,33 +22,46 @@
...
-    String url = getURI("/marccat/authority-record");
+  public void saveNameReturn201Status() throws IOException {
+    String url = getURI(AUTHORITY_RECORD_URL);
     Map<String, String> headers = addDefaultHeaders(url, StorageTestSuite.TENANT_ID);
     String templateJson = IOUtils.toString(this.getClass().getResourceAsStream("/authority/name.json"),
         String.valueOf(StandardCharsets.UTF_8));
 
-    given().headers("Content-Type", "application/json").headers(headers).queryParam("view", "-1")
-        .queryParam("lang", "eng").body(templateJson).when().post(url).then().statusCode(201);
+    given().headers(CONTENT_TYPE, FILE_TYPE).headers(headers).queryParam("view", "-1").queryParam("lang", "eng")
```
Comment (carmentrazza) [inline src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java:36]: Line up like other bibliographic tests

File: src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java [hunk 18]
```diff
@@ -22,33 +22,46 @@
...
 
-    String url = getURI("/marccat/authority-record");
+    String url = getURI(AUTHORITY_RECORD_URL);
     Map<String, String> headers = addDefaultHeaders(url, StorageTestSuite.TENANT_ID);
     String templateJson = IOUtils.toString(this.getClass().getResourceAsStream("/authority/name.json"),
         String.valueOf(StandardCharsets.UTF_8));
 
-    given().headers("Content-Type", "application/json").headers(headers).queryParam("view", "-1")
-        .queryParam("lang", "eng").body(templateJson).when().post(url).then().statusCode(201);
+    given().headers(CONTENT_TYPE, FILE_TYPE).headers(headers).queryParam("view", "-1").queryParam("lang", "eng")
```
Comment (carmentrazza) [inline src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java:49]: Line up like other bibliographic tests

File: src/main/java/org/folio/marccat/shared/CatalogingInformation.java [hunk 19]
```diff
@@ -52,57 +46,63 @@ static FixedField getFixedField(final StorageService storageService,
...
+        }
+        break;
+
+      case Global.OTHER_MATERIAL_TAG_CODE:
+        generalInformation = new GeneralInformation();
+        generalInformation.setDefaultValues(serviceConfiguration);
+
+        generalInformation.setHeaderType(headerTypeCode);
+        final Map<String, Object> mapRecordTypeMaterialHeader = storageService
+            .getMaterialTypeInfosByHeaderCode(headerTypeCode, code);
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/shared/CatalogingInformation.java:78]: Remember to format to 120 characters

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java (hunk 2 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java:12]: I have removed them.

File: src/main/java/org/folio/marccat/resources/AuthorityRecordAPI.java (hunk 3 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/resources/AuthorityRecordAPI.java:71]: I have changed it

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java (hunk 7 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:55]: I have changed it

File: src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java (hunk 9 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/resources/BibliographicRecordAPI.java:59]: I have changed it

File: src/main/java/org/folio/marccat/resources/shared/RecordUtils.java (hunk 14 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/resources/shared/RecordUtils.java:33]: I have changed it

File: src/main/java/org/folio/marccat/resources/shared/RecordUtils.java (hunk 15 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/resources/shared/RecordUtils.java:87]: I have changed it

File: src/main/resources/hibernate.cfg.xml (hunk 16 above)
Comment (eturienzo) [reply src/main/resources/hibernate.cfg.xml:11]: I have changed it

File: src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java (hunk 17 above)
Comment (eturienzo) [reply src/test/java/org/folio/marccat/resources/AuthorityRecordTest.java:36]: I have changed it

File: src/main/java/org/folio/marccat/shared/CatalogingInformation.java (hunk 19 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/shared/CatalogingInformation.java:78]: I have changed it

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (eturienzo) [reply descriptors/ModuleDescriptor-template.json:315]: I have add the permissions

File: src/main/java/org/folio/marccat/resources/BaseResource.java (hunk 4 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/resources/BaseResource.java]: I have add another class

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  /**
+   * Gets the authority model DAO.
+   *
+   * @return the authority model DAO
+   */
+  public AuthorityModelDAO getAuthorityModelDAO() {
+    return new AuthorityModelDAO();
+  }
+
+  @Deprecated
```
Comment (carmentrazza) [inline src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java:26]: Add the missing @deprecated Javadoc tag.

**Review by carmentrazza — APPROVED** [review]
Solve code smells in  AuthorityModel

File: src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java (hunk 20 above)
Comment (eturienzo) [reply src/main/java/org/folio/marccat/dao/persistence/AuthorityModel.java:29]: I have added

**Review by LeireDiez — APPROVED** [review]

**Pull request merged**